### PR TITLE
feat: support megatron-energon dataset

### PR DESCRIPTION
**Pull request by ziqi-wlb**
### What does this PR do?
Support load [Megatron-Energon](https://github.com/NVIDIA/Megatron-Energon) Dataset

> Megatron-Energon_ offers enhanced efficiency, flexibility, scalability and rich metadata support, making it an ideal choice for large-scale distributed training

### API and Usage Example

add **energon** datasets_type
- usage:
```sh
    # use megatron-energon type dataset
    --data.train_path $DATA_PATH \
    --data.datasets_type energon \
```

- example script:
```sh
MODEL_PATH=/cpfs/user/weishi/models/Qwen3-30B-A3B
SAVE_PATH=Qwen3-30B-A3B
CONFIG=configs/pretrain/qwen3-moe.yaml
DATA_PATH=/prodcpfs/user/weishi/data/text_data_converted/pile_test/webdataset
bash train.sh tasks/train_torch.py $CONFIG \
    --model.model_path ${MODEL_PATH} \
    --data.train_path $DATA_PATH \
    --data.num_workers 32 \
    --data.datasets_type energon \
    --train.global_batch_size 128 \
    --train.lr 5e-7 \
    --train.ulysses_parallel_size 1 \
    --train.save_steps 10 \
    --train.max_steps 20 \
    --train.ckpt_manager dcp \
    --train.output_dir ${SAVE_PATH}
```

- dataset logs:
image

### CODE REVIEW

**Comment by Fazziekey** [comment]
thanks for your MR

**Comment by Fazziekey** [comment]
can you make a git rebase?

**Comment by ziqi-wlb** [comment]
> can you make a git rebase?

done.

**Comment by Fazziekey** [comment]
add megatron-energon dependencies to pyproject.toml if possible? and run pre commit

**Comment by ziqi-wlb** [comment]
> add megatron-energon dependencies to pyproject.toml if possible? and run pre commit

done

File: pyproject.toml [hunk 1]
```diff
@@ -20,6 +20,7 @@ dependencies = [
   "blobfile>=3.0.0",
   "bytecheckpoint",
   "datasets>=2.16.0,<=2.21.0",
+  "megatron-energon>=7.2.1",
```
Comment (Luosuu) [inline pyproject.toml]: let's move it to optional instead?

Comment (Luosuu) [reply pyproject.toml]: my thought is that we only add absolute necessary package in the core dependencies

Comment (ziqi-wlb) [reply pyproject.toml]: done

**Comment by Luosuu** [comment]
Thank you! Is it possible for you to run a demo training and show the loss curve and then we can merge!

**Comment by ziqi-wlb** [comment]
> Thank you! Is it possible for you to run a demo training and show the loss curve and then we can merge!

image
qwen2.5-7B + small megatron-energon dataset (from pretrain bin/idx's test dataset)

**Comment by Fazziekey** [comment]
Thanks for you contribution，approved

**Review by Luosuu — APPROVED** [review]
